downsampler/src/main/scala/filodb.downsampler/PartitionDownsampler.scala

```diff
@@ -0,0 +1,106 @@
+package filodb.downsampler
+
+import scala.concurrent.duration.FiniteDuration
+
+import com.typesafe.scalalogging.StrictLogging
+import scalaxy.loops._
+
+import filodb.core.downsample._
+import filodb.core.memstore.{TimeSeriesPartition, TimeSeriesShardStats, WriteBufferPool}
+import filodb.core.metadata.Schema
+import filodb.core.store.{AllChunkScan, ReadablePartition}
+import filodb.memory.{BlockMemFactory, MemFactory}
+import filodb.memory.format.SeqRowReader
+
+object PartitionDownsampler extends StrictLogging {
+
+  /**
+    * Creates new donwsample partitions per per the resolutions
+    * * specified by `bufferPools`.
+    * Downsamples all chunks in `partToDownsample` per the resolutions and stoes
+    * downsampled data into the newly created partition.
+    *
+    * @param nativeMemManager downsampled chunks are stored into offheap memory
+    *                         allocated by this memory manager
+    * @param bufferPool buffers for the downsampled partitions are procured from this pool
+    *
+    * @return a TimeSeriesPartition for each resolution to be downsampled.
+    *         NOTE THAT THE PARTITIONS NEED TO BE FREED/SHUT DOWN ONCE CHUNKS ARE EXTRACTED FROM THEM
+    */
+  // scalastyle:off method.length parameter.number
+  def downsample(shardStats: TimeSeriesShardStats,
+                 resolutions: Seq[FiniteDuration],
+                 downsampleSchema: Schema,
+                 downsamplers: Seq[ChunkDownsampler],
+                 partToDownsample: ReadablePartition,
+                 blockHolder: BlockMemFactory,
+                 bufferPool: WriteBufferPool,
+                 nativeMemManager: MemFactory,
+                 downsampleIngestionTime: Long,
+                 userStartTime: Long,
+                 userEndTime: Long): Map[FiniteDuration, TimeSeriesPartition] = {
+
+    val resToPartition = resolutions.map { res =>
+      val part = new TimeSeriesPartition(
+        0,
+        downsampleSchema,
+        partToDownsample.partitionKey,
+        0,
+        bufferPool,
+        shardStats,
+        nativeMemManager,
+        1)
+      logger.trace(s"Creating new part=${part.hashCode}")
+      res -> part
+    }.toMap
+
+    val chunksets = partToDownsample.infos(AllChunkScan)
+    val timestampCol = 0
+    // TODO create a rowReader that will not box the vals below
```

Comment:
I think this needs to be fixed before scale testing.  Using an array of Any will box everything and create huge GC churn.

It's fairly simple to use `RecordBuilder` to achieve this.  Just call reset() every time so that there is only one record ever written.